heirloom-devtools/sccs/man/prt.1
.\"
.\" CDDL HEADER START
.\"
.\" The contents of this file are subject to the terms of the
.\" Common Development and Distribution License (the "License").  
.\" You may not use this file except in compliance with the License.
.\"
.\" You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
.\" or http://www.opensolaris.org/os/licensing.
.\" See the License for the specific language governing permissions
.\" and limitations under the License.
.\"
.\" When distributing Covered Code, include this CDDL HEADER in each
.\" file and include the License file at usr/src/OPENSOLARIS.LICENSE.
.\" If applicable, add the following below this CDDL HEADER, with the
.\" fields enclosed by brackets "[]" replaced with your own identifying
.\" information: Portions Copyright [yyyy] [name of copyright owner]
.\"
.\" CDDL HEADER END
.\" Copyright (c) 1990, Sun Microsystems, Inc.
.\"
.\" Portions Copyright (c) 2007 Gunnar Ritter, Freiburg i. Br., Germany
.\"
.\" Sccsid @(#)prt.1	1.7 (gritter) 02/01/07
.\"
.\" from OpenSolaris sccs-prt 1 "5 Oct 1990" "SunOS 5.11" "User Commands"
.TH PRT 1 "02/01/07" "Heirloom Development Tools" "User Commands"
.SH NAME
prt \- display delta table information from an SCCS file
.SH SYNOPSIS
.HP
.ad l
.nh
\fBprt\fR [\fB\-abdefistu\fR]
[\fB\-c\fR\fIdate-time\fR] [\fB\-r\fR\fIdate-time\fR] 
[\fB\-y\fR\fIsid\fR] \fIs.filename\fR...
.br
.ad b
.hy 1
.SH DESCRIPTION
.LP
\fIprt\fR prints selected portions of an \s-1SCCS\s+1 file.
By default, it prints the delta table (version log).
.PP
If a directory name is used in place of the \fIs.filename\fR argument, the \fIprt\fR command applies to all \fBs.\fRfiles in that directory.
Unreadable \fBs.\fRfiles produce an error; processing continues with the next file (if any).
The use of `\fB\(mi\fR' as the \fIs.filename\fR argument indicates that the names
of files are to be read from the standard input, one \fBs.\fRfile per line.
.PP
If any option other than \fB\-y\fR, \fB\-c\fR, or \fB\-r\fR is supplied, the name of each file being processed (preceded by one \fBNEWLINE\fR
and followed by two \fBNEWLINE\fR characters) appears above its contents.
.PP
If none of the \fB\-u\fR, \fB\-f\fR, \fB\-t\fR, or \fB\-b\fR options are used, \fB\-d\fR is assumed.
\fB\-s\fR, \fB\-i\fR are mutually
exclusive, as are \fB\-c\fR and \fB\-r\fR.
.TP
\fB\-a\fR
Display log entries for all deltas, including those marked as removed.
.TP
\fB\-b\fR
Print the body of the \fBs.\fRfile.
.TP
\fB\-d\fR
Print delta table entries.
This is the default.
.TP
\fB\-e\fR
Everything.
This option implies \fB\-d\fR, \fB\-i\fR, \fB\-u\fR, \fB\-f\fR, and \fB\-t\fR.
.TP
\fB\-f\fR
Print the flags of each named \fBs.\fRfile.
.TP
\fB\-i\fR
Print the serial numbers of included, excluded, and ignored deltas.
.TP
\fB\-s\fR
Print only the first line of the delta table entries; that is, only up to the statistics.
.TP
\fB\-t\fR
Print the descriptive text contained in the \fBs.\fRfile.
.TP
\fB\-u\fR
Print the user-names and/or numerical group \s-1ID\s+1R of users allowed to make deltas.
.TP
\fB\-c\fR\fIdate-time\fR
Exclude delta table entries that are specified cutoff date and time.
Each entry is printed as a single line,
preceded by the name of the \s-1SCCS\s+1 file.
This format (also produced by \fB\-r\fR \fB,\fR and \fB\-y\fR) makes it easy to sort multiple delta
tables in chronological order.
When both \fB\-y\fR and \fB\-c\fR, or \fB\-y\fR and \fB\-r\fR are supplied, \fIprt\fR stops printing when the first of the
two conditions is met.
.TP
\fB\-r\fR\fIdate-time\fR
Exclude delta table entries that are newer than the specified cutoff date and time.
.TP
\fB\-y\fR\fIsid\fR
Exclude delta table entries made prior to the \s-1SID\s+1 specified.
If no delta in the
table has the specified SID, the entire table is printed.
If no \s-1SID\s+1 is specified, the most recent delta is printed.
.SS Output Format
The following format is used to print those portions of the \fBs.\fRfile that are specified by the various options.
.TP 3
.PD 0
\(bu
\fBNEWLINE\fR
.TP
\(bu
Type of delta (\fBD\fR or \fBR\fR)
.TP
\(bu
\fBSPACE\fR
.TP
\(bu
\s-1SCCS\s+1 delta \s-1ID\s+1 (SID)
.TP
\(bu
\fBTAB\fR
.TP
\(bu
Date and time of creation in the form: \fIyy\fR\fB/\fR\fImm\fR\fB/\fR\fBdd\fR \fIhh\fR\fB/\fR\fImm\fR\fB/\fR\fIss\fR
.TP
\(bu
\fBSPACE\fR
.TP
\(bu
Username the delta's creator
.TP
\(bu
\fBTAB\fR
.TP
\(bu
Serial number of the delta
.TP
\(bu
\fBSPACE\fR
.TP
\(bu
Predecessor delta's serial number
.TP
\(bu
\fBTAB\fR
.TP
\(bu
Line-by-line change statistics in the form: \fIinserted\fR\fB/\fR\fIdeleted\fR\fB/\fR\fIunchanged\fR
.TP
\(bu
\fBNEWLINE\fR
.TP
\(bu
List of included deltas, followed by a \fBNEWLINE\fR (only if there were any such deltas and the \fB\-i\fR options was used)
.TP
\(bu
List of excluded deltas, followed by a \fBNEWLINE\fR (only if there were any such deltas and the \fB\-i\fR options was used)
.TP
\(bu
List of ignored deltas, followed by a \fBNEWLINE\fR (only if there were any such deltas and the \fB\-i\fR options was used)
.TP
\(bu
List of modification requests (MRs), followed by a \fBNEWLINE\fR (only if any \fBMR\fR numbers were supplied).
.TP
\(bu
Lines of the delta commentary (if any), followed by a \fBNEWLINE.\fR
.PD
.SH EXAMPLES
The following command:
.RS
.sp
\fB$ prt\fR \fB\-y\fR \fBprogram.c\fR
.sp
.RE
produces a one-line display of the delta table entry for the most recent version:
.LP
\fBs.program.c:  D 1.6   88/07/06 21:39:39 username   5 4 00159/00080/00636\fR.\|.\|.
.SH SEE ALSO
admin(1), 
delta(1), 
get(1), 
prs(1), 
what(1), 
sccs(1),
sccsfile(5)
.SH DIAGNOSTICS
Use the \s-1SCCS\s+1 \fIhelp\fR command for explanations (see 
\fIhelp\fR(1)).


ex-vi/Changes
Release ...
* The screen buffers for visual mode are now dynamically allocated, so
  vi usually does not return to ex mode with "screen too large" when the
  terminal is resized on a large monitor anymore.
* ex can now edit files with lines of arbitrary length. vi currently only
  shows the beginning of a line that does not fit onto the screen in its
  entirety.
* Viewing executables and compressed files is no longer inhibited.

Release 3/25/05
* vi no longer dies with a segmentation fault if a line does not fit on the
  screen after an insertion.
* The 'p' command now works correctly if the buffer contains a partial line
  with multibyte characters.
* Traditional regular expressions sometimes failed to operate correctly
  since the last release.

Release 2/25/05
* Traditional regular expressions can now be used with multibyte characters.
* When the 'ignorecase' option is toggled, saved regular expressions are now
  updated accordingly. (P)
* If a line began with a tabulator and another tabulator was inserted with
  the cursor located on the first tabulator, the display was not updated
  appropriately since the last release (Bugreport by Matthew Fischer). (P)
* Fixed a segmentation fault that occured in multibyte locales when operation
  was continued after vi had been stopped by ^Z, with the cursor positioned
  in a line longer than the terminal width as the last line of the current
  screen.
* Made multicolumn characters work in lines longer than the screen width in
  visual mode (Bugreport by Matthew Fischer).
* Made it work for Big5 locales (Patches by Matthew Fischer).
* Fixed a problem with the 'r' command in EUC-JP and Big5 locales (Bugreport
  by Matthew Fischer).
* The insertion of multicolumn characters and tab characters in multibyte
  locales now works with terminals that have the 'ic' but no 'im' termcap
  capability (Bugreport by Matthew Fischer).
* The argument to the -w option is correctly recognized now.
* If the SHELL environment variable is set to the empty string, it is now
  ignored.
* A non-null exit status is now returned if a file could not be opened, if
  an invalid address is given for a command, or if a tag is not found.
* If the match for a substitution is of length zero, a line overflow is
  now avoided when compiled with the 'UNIX(R) Regular Expression Library'.
* When moving left while the cursor is positioned over a multicolumn
  character at the end of the line, the bell is rung now (Bugreport by
  Matthew Fischer).
* When moving up or down to a row with different column arrangement while
  the cursor is positioned over a multicolumn character, the leftmost
  character above the original position is chosen in the new row.
* If both the -t and the -c option are given, the -t option is now
  processed first, i.e. the command is executed at the position where
  the tag was found.
* The -w option now also sets the scroll size for the 'z' command.
* When the name of a nonexisting file is given with the 'edit' ex command,
  the command now succeeds and prints a "[New file]" message.
* If standard output is not a terminal, no '\r' is written at the end of
  printed lines anymore.
* The 'source' ex command now works if command input comes from a pipe or
  regular file.
* Ex does not exit on errors immediately anymore if standard input is not
  a terminal but a pipe or regular file.
* The 'substitute' ex command can now be abbreviated as 'sub', 'subst' etc.
* A new version of mapmalloc.c that is derived from Unix 7th Edition code
  has been introduced.
* If the 'next!' ex command is given and the autowrite option is set, the
  current file is not written anymore.

Release 1/19/05
* The last release erroneously made 'X' work like 'x' in visual mode. It now
  deletes the character before the cursor again as documented (Bugreport by
  Matthew Fischer). (P)
* When a multicolumn character was replaced by another multicolumn character
  in insert mode, the display was not updated appropriately with terminals
  other than xterm.
* Various rendering errors happened with multicolumn characters if they
  started at an even column (counting from 1 upwards).
* When a multicolumn character was inserted and then replaced, the visual
  screen representation was sometimes not updated accordingly.
* Undoing the replacement of a multicolumn character by a singlecolumn
  character twice made the singlecolumn character invisible.
* The 'cw' visual command with a multibyte character as last character of
  the affected word located at the end of the line left garbage bytes past
  the end of the new text.
* Visual 'U' followed by 'u' lead to garbage on the screen when multibyte
  characters were present on the changed line.
* The position of '$' when changing text was fixed for cases where the first
  changed character had multiple columns but the last one had not.
* The handling of multicolumn characters was fixed for terminals without the
  IM (insert mode) capability. It is unlikely that such terminals actually
  exist, but vi will use the corresponding code for open mode if a termcap
  entry is not available.
* When an illegal multibyte sequence is entered in vi insert mode, no garbage
  bytes are generated anymore when the insert mode is left.
* The '\u', '\l', '\U', and '\L' substitution sequences work with multibyte
  characters now.
* Handle character case conversions with the '~' vi command correctly if the
  length of the converted multibyte sequence is smaller than the original one.
* Multibyte sequences that correspond to an unprintable character are now
  printed as multiple octal escape sequences.
* Pressing the ^U (kill) character in insert mode with a multibyte character
  at the beginning of an insertion at the start of a line sometimes positioned
  the cursor at weird locations since the last revision.
* Fixed an old vi bug: If a vi command that yanked or deleted part of a line
  was followed by an ex command that also yanked or deleted some text, a

heirloom-devtools/sccs/src/prt.c
/*
 * CDDL HEADER START
 *
 * The contents of this file are subject to the terms of the
 * Common Development and Distribution License (the "License").
 * You may not use this file except in compliance with the License.
 *
 * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
 * or http://www.opensolaris.org/os/licensing.
 * See the License for the specific language governing permissions
 * and limitations under the License.
 *
 * When distributing Covered Code, include this CDDL HEADER in each
 * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
 * If applicable, add the following below this CDDL HEADER, with the
 * fields enclosed by brackets "[]" replaced with your own identifying
 * information: Portions Copyright [yyyy] [name of copyright owner]
 *
 * CDDL HEADER END
 */
/* Copyright (c) 1984, 1986, 1987, 1988, 1989 AT&T */
/* All Rights Reserved */
/*
 * Copyright 2003 Sun Microsystems, Inc. All rights reserved.
 * Use is subject to license terms.
 */
/*
 * from prt.c 1.22 06/12/12
 */

/*
 * Copyright (c) 1980 Regents of the University of California.
 * All rights reserved. The Berkeley software License Agreement
 * specifies the terms and conditions for redistribution.
 */

/*	from OpenSolaris "ucbprt:prt.c"	*/

/*
 * Portions Copyright (c) 2006 Gunnar Ritter, Freiburg i. Br., Germany
 *
 * Sccsid @(#)prt.c	1.3 (gritter) 12/20/06
 */

/*
	Program to print parts or all of an SCCS file.
	Arguments to the program may appear in any order
	and consist of keyletters, which begin with '-',
	and named files.

	If a directory is given as an argument, each
	SCCS file within the directory is processed as if
	it had been specifically named. If a name of '-'
	is given, the standard input is read for a list
	of names of SCCS files to be processed.
	Non-SCCS files are ignored.
*/

# include	<defines.h>
# include	<had.h>
# include	<i18n.h>

# define NOEOF	0
# define BLANK(p)	while (!(*p == ' ' || *p == '\t')) p++;

char SccsError[MAXERRORLEN];
struct stat Statbuf;
static FILE *iptr;
static char *line = NULL;
static size_t line_size = 0;
static char statistics[25];
static struct delent {
	char type;
	char *osid;
	char *datetime;
	char *pgmr;
	char *serial;
	char *pred;
} del;
static int num_files;
static int prefix;
static time_t	cutoff;
static time_t	revcut;
static int linenum;
static char *ysid;
static char *flagdesc[26] = {
			NOGETTEXT(""),
			NOGETTEXT("branch"),
			NOGETTEXT("ceiling"),
			NOGETTEXT("default SID"),
			NOGETTEXT("encoded"),
			NOGETTEXT("floor"),
			NOGETTEXT(""),
			NOGETTEXT(""),
			NOGETTEXT("id keywd err/warn"),
			NOGETTEXT("joint edit"),
			NOGETTEXT(""),
			NOGETTEXT("locked releases"),
			NOGETTEXT("module"),
			NOGETTEXT("null delta"),
			NOGETTEXT(""),
			NOGETTEXT(""),
			NOGETTEXT("csect name"),
			NOGETTEXT(""),
			NOGETTEXT(""),
			NOGETTEXT("type"),
			NOGETTEXT(""),
			NOGETTEXT("validate MRs"),
			NOGETTEXT(""),
			NOGETTEXT(""),
			NOGETTEXT(""),
			NOGETTEXT("")
};

static void	prt(char *);
static void	getdel(register struct delent *, register char *);
static char	*read_to(int);
static char	*lineread(register int);
static void	printdel(register char *, register struct delent *);
static void	printit(register char *, register char *, register char *);

int 
main(int argc, char *argv[])
{
	register int j;
	register char *p;
	int  c;
	int testklt;
	extern int Fcnt;
	int current_optind;

	/*
	Set flags for 'fatal' to issue message, call clean-up
	routine, and terminate processing.
	*/
	Fflags = FTLMSG | FTLCLN | FTLEXIT;

	testklt = 1;

	/*
	The following loop processes keyletters and arguments.
	Note that these are processed only once for each
	invocation of 'main'.
	*/

	current_optind = 1;
	optind = 1;
	opterr = 0;
	j = 1;
	/*CONSTCOND*/
	while (1) {
			if(current_optind < optind) {
			   if (optind > j+1 ) {
			      if((argv[j+1][0] != '-') &&
				 ((argv[j][0] == 'c' && argv[j+1][0] <= '9') || 
				  (argv[j][0] == 'r' && argv[j+1][0] <= '9') ||
				  (argv[j][0] == 'y' && argv[j+1][0] <= '9'))) {
			        argv[j+1] = NULL;
			      } else {
				optind = j+1;
				current_optind = optind;
			      }
			      
			   }
			   if(argv[j][0] == '-') {
			     argv[j] = 0;
			   }
			   current_optind = optind;
			}
			j = current_optind;
		        c = getopt(argc, argv, "-r:c:y:esdaiuftbt");

				/* this takes care of options given after
				** file names.
				*/
			if((c == EOF)) {
			   if (optind < argc) {
				/* if it's due to -- then break; */
			       if(argv[j][0] == '-' &&
				      argv[j][1] == '-') {
			          argv[j] = 0;
			          break;
			       }
			       optind++;
			       current_optind = optind;
			       continue;
			   } else {
			       break;
			   }
			}
			p = optarg;
			switch (c) {
			case 'e':	/* print everything but body */
			case 's':	/* print only delta desc. and stats */
			case 'd':	/* print whole delta table */
			case 'a':	/* print all deltas */
			case 'i':	/* print inc, exc, and ignore info */
			case 'u':	/* print users allowed to do deltas */
			case 'f':	/* print flags */
			case 't':	/* print descriptive user-text */
			case 'b':	/* print body */
				break;
			case 'y':	/* delta cutoff */
				ysid = p;
				if(p[0] < '0'  || p[0] > '9') {
				  ysid = "";
				}
				prefix++;
				break;

			case 'c':	/* time cutoff */
				if (HADR) {
				   fatal("both 'c' and 'r' keyletters specified (pr2)");
				}
				if(p[0] ==  '-') {
				  /* no cutoff date given */
				  prefix++;
				  break;
				}
				if(p[0] > '9') {
				  prefix++;
				  break;
				}
				if (*p && parse_date(p,&cutoff))
					fatal("bad date/time (cm5)");
				prefix++;
				break;

			case 'r':	/* reverse time cutoff */
				if (HADC) {
				   fatal("both 'c' and 'r' keyletters specified (pr2)");
				}
				if(p[0] ==  '-') {
				  /* no cutoff date given */
				  prefix++;
				  break;
				}
				if(p[0] > '9') {
				  prefix++;
				  break;
				}
				if (*p && parse_date(p,&revcut))
					fatal ("bad date/time (cm5)");
				prefix++;
				break;

			default:
				fatal("Usage: prt [ -abdefistu ][ -c date-time ]\n\t[ -r date-time ][ -ySID ] s.filename...");
			}

			if (had[c - 'a']++ && testklt++)
				fatal("key letter twice (cm2)");
#if 0
			if(((c == 'c') || (c == 'r')||(c == 'y')) && (p[0] > '9')) {
			   argv[j] = NULL;
			   break;
			}
#endif
	}

	for(j=1; j<argc; j++){
		if(argv[j]) {
		       num_files++;
		}
	}

	if (num_files == 0)
		fatal("missing file arg (cm3)");

	if (HADC && HADR)
		fatal("both 'c' and 'r' keyletters specified (pr2)");

	setsig();

	/*
	Change flags for 'fatal' so that it will return to this
	routine (main) instead of terminating processing.
	*/
	Fflags &= ~FTLEXIT;
	Fflags |= FTLJMP;

	/*
	Call 'prt' routine for each file argument.
	*/
	for (j = 1; j < argc; j++)
		if ((p = argv[j]) != NULL)
			do_file(p, prt, 1);

	return (Fcnt ? 1 : 0);
}


/*
	Routine that actually performs the 'prt' functions.
*/

static void 
prt(char *file)
{
	int stopdel;
	int user, flag, text;
	char *p;
	time_t	bindate;
#if defined(BUG_1205145) || defined(GMT_TIME)
	time_t	tim;

	tzset();
#endif

	if (setjmp(Fjmp))	/* set up to return here from 'fatal' */
		return;		/* and return to caller of prt */

	if (HADE)
		HADD = HADI = HADU = HADF = HADT = 1;

	if (!HADU && !HADF && !HADT && !HADB)
		HADD = 1;

	if (!HADD)
		HADR = HADS = HADA = HADI = HADY = HADC = 0;

	if (HADS && HADI)
		fatal("s and i conflict (pr1)");

	iptr = xfopen(file,0);
	linenum = 0;

	p = lineread(NOEOF);
	if (*p++ != CTLCHAR || *p != HEAD)
		fatal("not an sccs file (co2)");

	stopdel = 0;

	if (!prefix) {
		printf("\n%s:\n",file);
	}
	if (HADD) {
		while (((p = lineread(NOEOF)) != NULL) && *p++ == CTLCHAR &&
				*p++ == STATS && !stopdel) {
			NONBLANK(p);
			copy(p,statistics);

			p = lineread(NOEOF);
			getdel(&del,p);
#if defined(BUG_1205145) || defined(GMT_TIME)
			date_ab(del.datetime,&bindate);
			/*
			Local time corrections before date_ba() call.
			Because this function uses gmtime() instead of localtime().
			*/
			tim = bindate - timezone;		/* GMT correction */
			if((localtime(&tim))->tm_isdst)
				tim += 1*60*60;			/* daylight savings */
			date_ba(&tim,del.datetime);
#endif

			if (!HADA && del.type != 'D') {
				read_to(EDELTAB);
				continue;
			}
			if (HADC) {
#if !(defined(BUG_1205145) || defined(GMT_TIME))
				date_ab(del.datetime,&bindate);
#endif
				if (bindate < cutoff) {
					stopdel = 1;
					break;
				}
			}
			if (HADR) {
#if !(defined(BUG_1205145) || defined(GMT_TIME))
				date_ab(del.datetime,&bindate);
#endif
				if (bindate >= revcut) {
					read_to(EDELTAB);
					continue;
				}
			}
			if (HADY && (equal(del.osid,ysid) || !(*ysid)))
				stopdel = 1;

			printdel(file,&del);

			while (((p = lineread(NOEOF)) != NULL) && *p++ == CTLCHAR) {
				if (*p == EDELTAB)
					break;
				switch (*p) {
				case INCLUDE:
					if (HADI)
						printit(file,"Included:\t",p);
					break;

				case EXCLUDE:
					if (HADI)
						printit(file,"Excluded:\t",p);
					break;

				case IGNORE:
					if (HADI)
						printit(file,"Ignored:\t",p);
					break;

				case MRNUM:
					if (!HADS)
						printit(file,"MRs:\t",p);
					break;

				case COMMENTS:
					if (!HADS)
						printit(file,"",p);
					break;

				default:
					sprintf(SccsError, "format error at line %d (co4)", linenum);
					fatal(SccsError);
				}
			}
		}
		if (prefix)
			printf("\n");

		if (stopdel && !(line[0] == CTLCHAR && line[1] == BUSERNAM))
			read_to(BUSERNAM);
	}
	else
		read_to(BUSERNAM);

	if (HADU) {
		user = 0;
		printf("\n Users allowed to make deltas -- \n");
		while (((p = lineread(NOEOF)) != NULL) && *p != CTLCHAR) {
			user = 1;
			printf("\t%s",p);
		}
		if (!user)
			printf("\teveryone\n");
	}
	else
		read_to(EUSERNAM);

	if (HADF) {
		flag = 0;
		printf("\n%s\n", "Flags --");
		while (((p = lineread(NOEOF)) != NULL) && *p++ == CTLCHAR &&
				*p++ == FLAG) {
			flag = 1;
			NONBLANK(p);
			/*
			 * The 'e' flag (file in encoded form) requires
			 * special treatment, as some versions of admin
			 * force the flag to be present (with operand 0)
			 * even when the user didn't explicitly specify
			 * it.  Stated differently, this flag has somewhat
			 * different semantics than the other binary-
			 * valued flags.
			 */
			if (*p == 'e') {
				/*
				 * Look for operand value; print description
				 * only if the operand value exists and is '1'.
				 */
				if (*++p) {
					NONBLANK(p);
					if (*p == '1')
						printf("\t%s",
							flagdesc['e' - 'a']);
				}
			}
			else {
				/*
				 * Standard flag: print description and
				 * operand value if present.
				 */
				printf("\t%s", flagdesc[*p - 'a']);

				if (*++p) {
					NONBLANK(p);
					printf("\t%s", p);
				}
			}
		}
		if (!flag)
			printf("\tnone\n");
	}
	else
		read_to(BUSERTXT);

	if (HADT) {
		text = 0;
		printf("\nDescription --\n");
		while (((p = lineread(NOEOF)) != NULL) && *p != CTLCHAR) {
			text = 1;
			printf("\t%s",p);
		}
		if (!text)
			printf("\tnone\n");
	}
	else
		read_to(EUSERTXT);

	if (HADB) {
		printf("\n");
		while (p = lineread(EOF))
			if (*p == CTLCHAR)
				printf("*** %s", ++p);
			else
				printf("\t%s", p);
	}

	fclose(iptr);
}


static void 
getdel(register struct delent *delp, register char *lp)
{
	lp += 2;
	NONBLANK(lp);
	delp->type = *lp++;
	NONBLANK(lp);
	delp->osid = lp;
	BLANK(lp);
	*lp++ = '\0';
	NONBLANK(lp);
	delp->datetime = lp;
	BLANK(lp);
	NONBLANK(lp);
	BLANK(lp);
	*lp++ = '\0';
	NONBLANK(lp);
	delp->pgmr = lp;
	BLANK(lp);
	*lp++ = '\0';
	NONBLANK(lp);
	delp->serial = lp;
	BLANK(lp);
	*lp++ = '\0';
	NONBLANK(lp);
	delp->pred = lp;
	repl(lp,'\n','\0');
}


static char *
read_to(int ch)
{
	char *n;

	while (((n = lineread(NOEOF)) != NULL) &&
			!(*n++ == CTLCHAR && *n == ch))
		continue;

	return(n);
}


static char *
lineread(register int eof)
{
	char	buf[512];
	size_t	read, used = 0;

	do {
		if (fgets(buf, sizeof(buf), iptr) == NULL) {
			if (!eof) {
				fatal("premature eof (co5)");
			}
			return (used) ? line : NULL;
		}

		read = strlen(buf);

		if ((used + read) >= line_size) {
			line_size += sizeof(buf);
			line = (char*) realloc(line, line_size);
			if (line == NULL) {
				fatal("OUT OF SPACE (ut9)");
			}
		}

		strcpy(line + used, buf);
		used += read;
	} while (line[used - 1] != '\n');

	linenum++;

	return line;
}


static void 
printdel(register char *file, register struct delent *delp)
{
	printf("\n");

	if (prefix) {
		statistics[length(statistics) - 1] = '\0';
		printf("%s:\t",file);
	}

	printf("%c %s\t%s %s\t%s %s\t%s",delp->type,delp->osid,
		delp->datetime,delp->pgmr,delp->serial,delp->pred,statistics);
}


/*ARGSUSED*/
static void 
printit(register char *file, register char *str, register char *cp)
{
	cp++;
	NONBLANK(cp);

	if (prefix) {
		cp[length(cp) - 1] = '\0';
		printf(" ");
	}

	printf("%s%s",str,cp);
}

/* for fatal() */
void 
clean_up(void)
{
}
